reporting/performance/test_mode.cpp

```diff
@@ -0,0 +1,39 @@
+//  (C) Copyright Nick Thompson 2020.
+//  Use, modification and distribution are subject to the
+//  Boost Software License, Version 1.0. (See accompanying file
+//  LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <random>
+#include <boost/math/statistics/univariate_statistics.hpp>
+#include <boost/multiprecision/cpp_bin_float.hpp>
+#include <benchmark/benchmark.h>
+
+using boost::multiprecision::cpp_bin_float_50;
+
+template <class Z>
+void test_mode(benchmark::State& state)
+{
+    using boost::math::statistics::mode;
+    
+    std::random_device rd;
+    std::mt19937_64 mt(rd());
+    std::uniform_real_distribution<double> dist {1, 10};
+
+    auto gen = [&dist, &mt](){return dist(mt);};
+
+    std::vector<Z> v(100);
+    std::generate(v.begin(), v.end(), gen);
+
+    for (auto _ : state)
+    {
+        benchmark::DoNotOptimize(mode(v));
+    }
+}
+
+BENCHMARK_TEMPLATE(test_mode, float);
+BENCHMARK_TEMPLATE(test_mode, double);
+BENCHMARK_TEMPLATE(test_mode, long double);
+BENCHMARK_TEMPLATE(test_mode, int);
```

Comment:
I would try int32_t, int64_t, uint64_t, uint32_t, and maybe __int128_t.

Then we need to measure the complexity. My apologies for recommending `rsqrt_performance.cpp` as a template, [chebyshev_clenshaw.cpp](https://github.com/boostorg/math/blob/develop/reporting/performance/chebyshev_clenshaw.cpp) is a better example for this.

Could you also add a benchmark for the vector {0,1,2,3,4,5,6,...}  and {1,1,2,2,3,3,4,4, ...}? I want to see if there's some way to break the complexity of the algorithm.

Reply:
Added those additional tests. Attached is a synopsis of the results.

[Results.txt](https://github.com/boostorg/math/files/4873188/Results.txt)